### PR TITLE
[SPARK-47307][SQL] Add a config to optionally chunk base64 strings

### PR DESCRIPTION
**Pull request by wForget**
Follow up #45408

### What changes were proposed in this pull request?
[[SPARK-47307](https://issues.apache.org/jira/browse/SPARK-47307)] Add a config to optionally chunk base64 strings

### Why are the changes needed?
In #35110, it was incorrectly asserted that:

> ApacheCommonBase64 obeys http://www.ietf.org/rfc/rfc2045.txt

This is not true as the previous code called:

```java
public static byte[] encodeBase64(byte[] binaryData)
```

Which states:

> Encodes binary data using the base64 algorithm but does not chunk the output.

However, the RFC 2045 (MIME) base64 encoder does chunk by default. This now means that any Spark encoded base64 strings cannot be decoded by encoders that do not implement RFC 2045. The docs state RFC 4648.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing test suite.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 1]
```diff
@@ -2685,18 +2685,32 @@ case class Chr(child: Expression)
 case class Base64(child: Expression)
   extends UnaryExpression with ImplicitCastInputTypes with NullIntolerant {
 
+  lazy val chunkBase64: Boolean = SQLConf.get.chunkBase64StringEnabled
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:2688]: let's move it to the constructor

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 2]
```diff
@@ -2685,18 +2685,32 @@ case class Chr(child: Expression)
...
   }
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     nullSafeCodeGen(ctx, ev, (child) => {
-      s"""${ev.value} = UTF8String.fromBytes(
-            ${classOf[JBase64].getName}.getMimeEncoder().encode($child));
-       """})
+      if (chunkBase64) {
+        s"""${ev.value} = UTF8String.fromBytes(
+             ${classOf[JBase64].getName}.getMimeEncoder().encode($child));
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:2706]: Why don't we use the encoder directly?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -3525,6 +3525,13 @@ object SQLConf {
     .booleanConf
     .createWithDefault(false)
 
+  val CHUNK_BASE_64_STRING_ENABLED = buildConf("spark.sql.chunkBase64String.enabled")
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3528]: We shall make scope in the desc more specific a a lot of other places are using base64 too

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -3525,6 +3525,13 @@ object SQLConf {
...
+    .doc("When true, base64 strings generated by the base64 function are chunked into lines of " +
+      "at most 76 characters. When false, the base64 strings are not chunked.")
+    .version("4.0.0")
+    .booleanConf
+    .createWithDefault(true)
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3533]: We will probably make it false by default and an legacy internal configuration

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 2 above)
Comment (wForget) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:2706]: > Why don't we use the encoder directly?

`java.util.Base64$Encoder` is not serializable.

```
-- !query
select base64(c7), base64(c8), base64(v), ascii(s) from char_tbl4
-- !query schema
struct<>
-- !query output
java.io.NotSerializableException
java.util.Base64$Encoder
Serialization stack:
	- object not serializable (class: java.util.Base64$Encoder, value: java.util.Base64$Encoder@423ed07f)
	- element of array (index: 2)
	- array (class [Ljava.lang.Object;, size 5)
	- field (class: org.apache.spark.sql.execution.WholeStageCodegenEvaluatorFactory, name: org$apache$spark$sql$execution$WholeStageCodegenEvaluatorFactory$$references, type: class [Ljava.lang.Object;)
	- object (class org.apache.spark.sql.execution.WholeStageCodegenEvaluatorFactory, org.apache.spark.sql.execution.WholeStageCodegenEvaluatorFactory@2fd9633e)
	- element of array (index: 0)
	- array (class [Ljava.lang.Object;, size 1)
	- field (class: java.lang.invoke.SerializedLambda, name: capturedArgs, type: class [Ljava.lang.Object;)
	- object (class java.lang.invoke.SerializedLambda, SerializedLambda[capturingClass=class org.apache.spark.sql.execution.WholeStageCodegenExec, functionalInterfaceMethod=scala/Function2.apply:(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;, implementation=invokeStatic org/apache/spark/sql/execution/WholeStageCodegenExec.$anonfun$doExecute$4$adapted:(Lorg/apache/spark/sql/execution/WholeStageCodegenEvaluatorFactory;Ljava/lang/Object;Lscala/collection/Iterator;)Lscala/collection/Iterator;, instantiatedMethodType=(Ljava/lang/Object;Lscala/collection/Iterator;)Lscala/collection/Iterator;, numCaptured=1])
	- writeReplace data (class: java.lang.invoke.SerializedLambda)
	- object (class org.apache.spark.sql.execution.WholeStageCodegenExec$$Lambda$2458/0x000002cf3e949c30, org.apache.spark.sql.execution.WholeStageCodegenExec$$Lambda$2458/0x000002cf3e949c30@603a0fa7)

```

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:2707]: maybe we can follow the work I have done for `Encode` to make it RuntimeReplaceable with StaticInvoke

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 5]
```diff
@@ -3525,6 +3525,14 @@ object SQLConf {
     .booleanConf
     .createWithDefault(false)
 
+  val CHUNK_BASE_64_STRING_ENABLED = buildConf("spark.sql.chunkBase64String.enabled")
+    .doc("Whether to truncate string generated by the `Base64` function. When true, base64" +
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3529]:
```suggestion
  val CHUNK_BASE_64_STRING_ENABLED = buildConf("spark.sql.legacy.chunkBase64String.enabled")
    .internal()
    .doc("Whether to truncate string generated by the `Base64` function. When true, base64" +
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 2 above)
Comment (wForget) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:2707]: > maybe we can follow the work I have done for `Encode` to make it RuntimeReplaceable with StaticInvoke

Thanks, I will try it

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 6]
```diff
@@ -3525,6 +3525,15 @@ object SQLConf {
     .booleanConf
     .createWithDefault(false)
 
+  val CHUNK_BASE_64_STRING_ENABLED = buildConf("spark.sql.legacy.chunkBase64String.enabled")
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3528]: nit: BASE_64 to BASE64

File: connect/common/src/test/resources/query-tests/explain-results/function_base64.explain [hunk 7]
```diff
@@ -1,2 +1,2 @@
-Project [base64(cast(g#0 as binary)) AS base64(CAST(g AS BINARY))#0]
+Project [static_invoke(Base64.encode(cast(g#0 as binary), false)) AS base64(CAST(g AS BINARY))#0]
```
Comment (yaooqinn) [inline connect/common/src/test/resources/query-tests/explain-results/function_base64.explain:1]: we can omit `false` here

File: sql/core/src/test/resources/sql-tests/analyzer-results/charvarchar.sql.out [hunk 8]
```diff
@@ -465,7 +465,7 @@ Project [ascii(c7#x) AS ascii(c7)#x, ascii(c8#x) AS ascii(c8)#x, ascii(v#x) AS a
 -- !query
 select base64(c7), base64(c8), base64(v), ascii(s) from char_tbl4
 -- !query analysis
-Project [base64(cast(c7#x as binary)) AS base64(c7)#x, base64(cast(c8#x as binary)) AS base64(c8)#x, base64(cast(v#x as binary)) AS base64(v)#x, ascii(s#x) AS ascii(s)#x]
+Project [base64(cast(c7#x as binary), false) AS base64(c7)#x, base64(cast(c8#x as binary), false) AS base64(c8)#x, base64(cast(v#x as binary), false) AS base64(v)#x, ascii(s#x) AS ascii(s)#x]
```
Comment (yaooqinn) [inline sql/core/src/test/resources/sql-tests/analyzer-results/charvarchar.sql.out:468]: ditto

Comment (wForget) [reply sql/core/src/test/resources/sql-tests/analyzer-results/charvarchar.sql.out:468]: Do you mean to override `toString` method?

File: connect/common/src/test/resources/query-tests/explain-results/function_base64.explain (hunk 7 above)
Comment (wForget) [reply connect/common/src/test/resources/query-tests/explain-results/function_base64.explain:1]: How do I omit it? `false` also seems to be preserved in `function_encode.explain`.

https://github.com/apache/spark/blob/4c7edd2a20480a8521fcc88a966b22619143aebd/connect/common/src/test/resources/query-tests/explain-results/function_encode.explain#L1

**Comment by wForget** [comment]
I noticed a failed unit test in https://github.com/wForget/spark/actions/runs/9903281687/job/27358692534, should we add `org.apache.spark.sql.catalyst.expressions.Base64` to the `ignoreSet`?

```
 - check outputs of expression examples (10 seconds, 991 milliseconds)
[info] - Check whether SQL expressions should extend NullIntolerant *** FAILED *** (27 milliseconds)
[info]   org.apache.spark.sql.catalyst.expressions.Base64 should not extend NullIntolerant, or add org.apache.spark.sql.catalyst.expressions.Base64 in the ignoreSet of this test. (ExpressionInfoSuite.scala:290)
[info]   org.scalatest.exceptions.TestFailedException:
```

File: connect/common/src/test/resources/query-tests/explain-results/function_base64.explain (hunk 7 above)
Comment (yaooqinn) [reply connect/common/src/test/resources/query-tests/explain-results/function_base64.explain:1]: override toString

**Comment by yaooqinn** [comment]
UnaryExpression can be removed from inheritance to pass the test

**Review by yaooqinn — APPROVED** [review]
LGTM

File: connect/common/src/test/resources/query-tests/explain-results/function_base64.explain (hunk 7 above)
Comment (yaooqinn) [reply connect/common/src/test/resources/query-tests/explain-results/function_base64.explain:1]: I have found some other functions also having such issue, we can fix them together later

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 9]
```diff
@@ -3525,6 +3525,15 @@ object SQLConf {
     .booleanConf
     .createWithDefault(false)
 
+  val CHUNK_BASE64_STRING_ENABLED = buildConf("spark.sql.legacy.chunkBase64String.enabled")
+    .internal()
+    .doc("Whether to truncate string generated by the `Base64` function. When true, base64" +
+      " strings generated by the base64 function are chunked into lines of at most 76" +
+      " characters. When false, the base64 strings are not chunked.")
+    .version("4.0.0")
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3533]:
```suggestion
    .version("3.5.2")
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 10]
```diff
@@ -3525,6 +3525,15 @@ object SQLConf {
     .booleanConf
     .createWithDefault(false)
 
+  val CHUNK_BASE64_STRING_ENABLED = buildConf("spark.sql.legacy.chunkBase64String.enabled")
+    .internal()
+    .doc("Whether to truncate string generated by the `Base64` function. When true, base64" +
+      " strings generated by the base64 function are chunked into lines of at most 76" +
+      " characters. When false, the base64 strings are not chunked.")
+    .version("3.5.2")
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3533]: I'm going to pick this to 3.5.2

**Comment by yaooqinn** [comment]
Merged to master.

Thank you @wForget, and also @ted-jenks for the previous work. All credits are logged.

**Comment by yaooqinn** [comment]
It conflicts with branch-3.5. Please raise another backport PR when you have some time. Thank you in advance. - @wForget

**Comment by dongjoon-hyun** [comment]
cc @cloud-fan , too

**Comment by gatorsmile** [comment]
image

@wForget  is it true? 

@yaooqinn @dongjoon-hyun @wForget  I think we need to discuss it in the dev list before merging this PR. This will break the encryption that relies on base64. 

I would suggest reverting it first before making a common decision on the dev list.

**Comment by dongjoon-hyun** [comment]
Thank you for the feedback, @gatorsmile. 

I have three questions to understand your request.

- Do you see the example, [SPARK-47307](https://issues.apache.org/jira/browse/SPARK-47307)? What is your opinion?
- We have the following migration guide, do you want to set `spark.sql.legacy.chunkBase64String.enabled=true` by default in Apache Spark 3.5.2?

> Since 3.5.2, the `base64` function will return a non-chunked string. To restore the behavior of chunking base64 encoded strings into lines of at most 76 characters, set `spark.sql.legacy.chunkBase64String.enabled` to `true`.

- If we change the default value, do we still need to revert?

**Comment by cloud-fan** [comment]
I think the PR description should be improved to mention more information. It's a behavior change, we should clearly define the behavior difference and measure the impact. Does it mean the Spark base64 result can be decoded by more encoders now? Is there any standard encoder that may fail to decode after this change?

**Comment by wForget** [comment]
Sorry, I didn't change PR description after changing default value for new config.
Actually, a behavior change was unexpectedly introduced in #35110, this PR aims to make the behavior of base64 encode configurable. If the change in current behavior is controversial, I would first submit a PR to change default value of the newly introduced config which would revert to the previous behavior. WDYT?

**Comment by yaooqinn** [comment]
>  I would first submit a PR to change default value of the newly introduced config which would revert to the previous behavior. 
+1

**Pull request closed without merging**